File: app/src/main/java/com/demo/justapp/exchanger/presentation/rates/adapter/util/CurrencyRateAdapterDiffUtil.java
package com.demo.justapp.exchanger.presentation.rates.adapter.util;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.util.DiffUtil;

import com.demo.justapp.exchanger.models.local.CurrencyRate;

import java.util.List;

/**
 * @author Sergey Rodionov
 */
public class CurrencyRateAdapterDiffUtil extends DiffUtil.Callback {

    public static final String KEY_CURRENCY = "currency";
    public static final String KEY_AMOUNT = "amount";
    private List<CurrencyRate> mOldList;
    private List<CurrencyRate> mNewList;

    /**
     * @param oldList старый лист с данными
     * @param newList новый лист с данными
     */
    public CurrencyRateAdapterDiffUtil(@NonNull List<CurrencyRate> oldList,
                                       @NonNull List<CurrencyRate> newList) {
        mOldList = oldList;
        mNewList = newList;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int getOldListSize() {
        return mOldList.size();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int getNewListSize() {
        return mNewList.size();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return mOldList.get(oldItemPosition) == mNewList.get(newItemPosition);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        CurrencyRate oldItem = mOldList.get(oldItemPosition);
        CurrencyRate newItem = mNewList.get(newItemPosition);

        return oldItem.equals(newItem);
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        CurrencyRate oldItem = mOldList.get(oldItemPosition);
        CurrencyRate newItem = mNewList.get(newItemPosition);

        Bundle diffBundle = new Bundle();
        if (!newItem.getCurrency().equals(oldItem.getCurrency())) {
            diffBundle.putString(KEY_CURRENCY, newItem.getCurrency());
        }
        if (newItem.getRateExchange() != oldItem.getRateExchange()) {
            diffBundle.putDouble(KEY_AMOUNT, oldItem.getRateExchange());
        }
        if (diffBundle.size() == 0) return null;
        return diffBundle;
    }
}

File: app/src/main/java/com/demo/justapp/exchanger/presentation/rates/adapter/util/CurrencyRateAdapterDiffUtil.kt
package com.demo.justapp.exchanger.presentation.rates.adapter.util

import android.os.Bundle
import android.support.v7.util.DiffUtil
import com.demo.justapp.exchanger.models.local.CurrencyRate

/**
 * @author Sergey Rodionov
 */
class CurrencyRateAdapterDiffUtil(oldList: List<CurrencyRate>,
                                  newList: List<CurrencyRate>)
    : DiffUtil.Callback() {

    companion object {
        internal const val KEY_CURRENCY = "currency"
        internal const val KEY_AMOUNT = "amount"
    }

    private val mOldList: List<CurrencyRate> = oldList
    private val mNewList: List<CurrencyRate> = newList

    override fun getOldListSize(): Int {
        return mOldList.size
    }

    override fun getNewListSize(): Int {
        return mNewList.size
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return mOldList[oldItemPosition] == mNewList[newItemPosition]
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldCurrencyItem = mOldList[oldItemPosition]
        val newCurrencyItem = mNewList[newItemPosition]
        return oldCurrencyItem.equals(newCurrencyItem)
    }

    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
        val oldItem = mOldList[oldItemPosition]
        val newItem = mNewList[newItemPosition]
        val difBundle = Bundle()

        if (newItem.currency != oldItem.currency) {
            difBundle.putString(KEY_CURRENCY, newItem.currency)
        }

        if (newItem.rate != oldItem.rate) {
            difBundle.putDouble(KEY_AMOUNT, newItem.rate)
        }

        if (difBundle.size() == 0) {
            return null
        }

        return difBundle
    }

}